### PR TITLE
R2RDump - Fix errors caused by ExportedTypes rids in AvailableTypes section

### PR DESCRIPTION
**Pull request by acmyu**

### CODE REVIEW

**Comment by zamont** [comment]
Can you describe in greater detail what error is being fixed by this PR?

**Comment by acmyu** [comment]
The AvailableTypes section can contain rids to both TypeDefs and ExportedTypes. If it's an ExportedType rid and it uses it to get a TypeDef, it would cause an error. This pr uses the rid to get ExportedType if TypeDef fails, because the section doesn't contain info on whether it's TypeDef or ExportedType

File: src/tools/r2rdump/R2RReader.cs [hunk 1]
```diff
@@ -364,6 +364,9 @@ private void ParseAvailableTypes()
             {
                 return;
             }
+
+            Dictionary<uint, bool> added = new Dictionary<uint, bool>();
```
Comment (nattress) [inline src/tools/r2rdump/R2RReader.cs]: If you just want to quickly determine existence, a `HashSet<uint>` would be sufficient

File: src/tools/r2rdump/R2RReader.cs [hunk 2]
```diff
@@ -514,17 +537,46 @@ public int GetOffset(int rva)
...
+            string typeNamespace = "";
             string typeStr = "";
             do
             {
-                typeDef = mdReader.GetTypeDefinition(handle);
-                typeStr = "." + mdReader.GetString(typeDef.Name) + typeStr;
-                handle = typeDef.GetDeclaringType();
+                try
+                {
+                    TypeDefinition typeDef = mdReader.GetTypeDefinition(handle);
```
Comment (nattress) [inline src/tools/r2rdump/R2RReader.cs]: Just checking, do `GetTypeDefFullName` and `GetExportedTypeFullName` handle nested types properly?

Comment (acmyu) [reply src/tools/r2rdump/R2RReader.cs]: Do exported types have nested types? The ExportedType class doesn't provide GetDeclaringType() for getting the parent type.

GetTypeDefFullName does handle nested types. I'm not sure if there could be cases where TypeDefs are nested inside an exported type. That could cause an error

Comment (nattress) [reply src/tools/r2rdump/R2RReader.cs]: Yes I think you're right. Looks good to me.

**Comment by acmyu** [comment]
@dotnet-bot test Ubuntu x64 Formatting please

**Pull request merged**